---
title: "euk_asvs_mgns_regressions"
output: html_document
date: "2024-06-13"
---

### Library Loading
```{r}
library(here)
library(tidyverse)
library(RColorBrewer)
library(ggplot2)
library(purrr)
library(Polychrome)
```

### Load  data
```{r}
taxa <- read.csv(file=here("18S Comparison","Phylum_Class_18S_samplepairs.csv"))
```

### Group data by Phylum, sum up read and metagenome counts for each Class
```{r}
# returns tibble table
taxa_phy <- taxa %>% group_by(Phylum, Metagenome.sample, ASV.sample) %>% 
  summarise(mgn_sum=sum(percent_ORFs),
            reads_sum=sum(X18S.reads),
            .groups = 'drop')

# remove the "unknown" and "nan" phyla
taxa_phy <- taxa_phy[!taxa_phy$Phylum == 'nan', ]
taxa_phy <- taxa_phy[!taxa_phy$Phylum == 'unknown', ]

# change "Dinophyceae" to "Dinophyta"
taxa_phy$Phylum[taxa_phy$Phylum == "Dinophyceae"] <- "Dinophyta"

# subset only for the most abundant taxa to limit number of phyla to display (full number is 97 for COI, 98 for 18S)
taxa_phy_filt <- taxa_phy[taxa_phy$mgn_sum > 0.005 |
                            taxa_phy$reads_sum > 0.005, ]
```

### Scatter plot of all Phyla
```{r}
# create color palette
P50 = createPalette(55, c("#ff0000", "#00ff00", "#0000ff"))
P50 <- sortByHue(P50)
P50 <- as.vector(t(matrix(P50)))
names(P50) <- NULL

# reorder df in descending order of amplicon read frequencies
taxa_phy_plot <- taxa_phy_filt[order(-taxa_phy_filt$reads_sum), ]
phy_unique <- taxa_phy_plot[!duplicated(taxa_phy_plot$Phylum), ]
phy_unique_ordered <- phy_unique$Phylum[order(match(phy_unique$Phylum,
                                                  taxa_phy_plot$Phylum))] 
taxa_phy_plot$Phylum <- factor(taxa_phy_plot$Phylum, 
                               levels=phy_unique_ordered)
  
p <- ggplot(taxa_phy_plot, aes(x=reads_sum, y=mgn_sum, color=Phylum)) +
  geom_point(show.legend=TRUE) +
  labs(title = paste("Eukaryotic phyla in paired metagenome & 18S samples"),
         x="Fraction of 18S reads", y="Fraction of eukaryotic metagenome ORFs") +
  coord_cartesian(xlim = c(0, max(taxa_phy$reads_sum) + (0.1*max(taxa_phy$reads_sum))),
                    ylim = c(0, max(taxa_phy$mgn_sum) + 0.1*max(taxa_phy$mgn_sum))) +
  scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
  labs(color="Phylum") +
  theme_classic() +
  scale_color_manual(values=P50)

ggsave(here("18S Comparison", "Euk_phyla_18S_comparison.png"),
       p,width=12, height=6, dpi=300, units="in")
```

### Assess each Phylum

```{r}
# Split the data into subsets based on Phylum
data_splits <- split(taxa_phy_filt, taxa_phy_filt$Phylum)

# Initialize a list to store the linear regression models and R-squared values
lm_models <- list()
r_squared_values <- vector("numeric", length(data_splits))
reads_max <- vector("numeric", length(data_splits))

# Loop through the subsets and fit linear regression models
for (i in seq_along(data_splits)) {
  subset_data <- data_splits[[i]]
  lm_models[[i]] <- lm(reads_sum ~ mgn_sum, data = subset_data)
  r_squared_values[i] <- summary(lm_models[[i]])$r.squared
  reads_max[i] <- max(subset_data$reads_sum)
}
```

### If grouping taxa by r-squared values, run the cell below

```{r}
# Determine thresholds for creating subsets based on R-squared values
threshold1 <- 0.01
threshold2 <- 0.1
threshold3 <- 0.5

# Identify subsets with R-squared values above the threshold
subset1 <- which(r_squared_values < threshold1)
subset2 <- which(r_squared_values >= threshold1 & r_squared_values < threshold2)
subset3 <- which(r_squared_values >= threshold2 & r_squared_values < threshold3)
subset4 <- which(r_squared_values >= threshold3)
```

### If grouping taxa by max amplicon read fraction, run the cell below

```{r}
# Determine thresholds for creating subsets based on max amplicon read fraction
# COI
#threshold1 <- 0.005
#threshold2 <- 0.01
#threshold3 <- 0.1
# 18S
threshold1 <- 0.0001
threshold2 <- 0.01
threshold3 <- 0.1

# Identify subsets with max reads_sum values above the threshold
subset1 <- which(reads_max < threshold1)
subset2 <- which(reads_max >= threshold1 & reads_max < threshold2)
subset3 <- which(reads_max >= threshold2 & reads_max < threshold3)
subset4 <- which(reads_max >= threshold3)
```


```{r}
# Loop through the selected subsets and create plots
for (subset_index in subset1) {
  subset_data <- data_splits[[subset_index]]
  
  r_squared <- summary(lm_models[[subset_index]] )$r.squared
  slope <- summary(lm_models[[subset_index]] )$coef[[2]]
  
  # Create a scatter plot for the subset
  scatter_plot <- ggplot(subset_data, aes(x=reads_sum, y=mgn_sum)) +
    geom_point(color="#0072B2") +
    geom_smooth(method="lm", se=TRUE, color="black", linewidth=0.5) +
    labs(title = paste("Phylum: ", subset_data$Phylum),
         x="Fraction of 18S reads", y="Fraction of eukaryotic metagenome ORFs") +
    annotate(geom="text", label = paste("R-squared = ", round(r_squared, 2)),
             x=max(subset_data$reads_sum) - (0.2*max(subset_data$reads_sum)), 
              y=max(subset_data$mgn_sum) - 0.5*max(subset_data$mgn_sum),
              color="black", hjust=0, vjust=1) +
    annotate(geom="text", label = paste("Slope = ", round(slope, 2)),
             x=max(subset_data$reads_sum) - (0.2*max(subset_data$reads_sum)), 
              y=max(subset_data$mgn_sum) - 0.55*max(subset_data$mgn_sum),
              color="black", hjust=0, vjust=1) +
    coord_cartesian(xlim = c(0, max(subset_data$reads_sum) +
                               (0.1*max(subset_data$reads_sum))), 
                    ylim = c(0, max(subset_data$mgn_sum) + 0.1*max(subset_data$mgn_sum))) +
    scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
    theme_classic()

  # Print the scatter plot
  #print(scatter_plot)
  
  # Generate a unique filename
  filename <- paste("18Splot_", subset_data$Phylum, ".png", sep = "")
  
  # Save the plot
  ggsave(here("18S Comparison", "Phylum plots mmseqs", "subset_1", filename), scatter_plot, width = 7, height = 5, dpi = 300, units = "in")
}
```

Plot individual phyla with regression line
```{r}
data <- data_splits[['Streptophyta']]

# Generate regression model and calculate slope and r-squared values  
lm <- lm(reads_sum ~ mgn_sum, data = data)
r_squared <- summary(lm)$r.squared
slope <- lm$coef[[2]]

# with linear regression
p <- ggplot(data, aes(x=reads_sum, y=mgn_sum)) +
  geom_point(show.legend=TRUE, size=2, color="#0072B2") +
  geom_smooth(method="lm", se=TRUE, color="black", linewidth=0.5) +
  labs(title = paste("Phylum Streptophyta in paired metagenome & 18S samples"),
         x="Fraction of 18S reads", y="Fraction of eukaryotic metagenome ORFs") +
  annotate(geom="text", label = paste("R-squared = ", round(r_squared, 2)),
             x=max(data$reads_sum) - (0.2*max(data$reads_sum)), 
              y=max(data$mgn_sum) - 0.5*max(data$mgn_sum),
              color="black", hjust=0, vjust=1) +
  annotate(geom="text", label = paste("Slope = ", round(slope, 2)),
             x=max(data$reads_sum) - (0.2*max(data$reads_sum)), 
              y=max(data$mgn_sum) - 0.55*max(data$mgn_sum),
              color="black", hjust=0, vjust=1) +
  coord_cartesian(xlim = c(0, max(data$reads_sum) + (0.1*max(data$reads_sum))),
                    ylim = c(0, max(data$mgn_sum) + 0.1*max(data$mgn_sum))) +
  scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
  theme_classic() 

ggsave(here("18S Comparison", "Class plots mmseqs", "Bacillariophyta_18S_lm-slope-r2-se.png"), p,
      width=11, height=8, dpi=300, units="in")
```


Plot individual phyla colored by Class
```{r}
P6 = createPalette(24, c("#ff9933", "#00cc00", "#0080ff"))
P6 <- sortByHue(P6)
P11 <- as.vector(t(matrix(4)))
names(P6) <- NULL

r_squared <- summary(lm_models[[subset_index]] )$r.squared

phylum <- taxa[taxa$Phylum == 'Streptophyta', ]

# no linear regression
p <- ggplot(phylum, aes(x=X18S.reads, y=percent_ORFs, color=Class)) +
  geom_point(show.legend=TRUE, size=2) +
  labs(title = paste("Streptophyta classes in paired metagenome & COI samples"),
         x="Fraction of 18S reads", y="Fraction of eukaryotic metagenome ORFs") +
  coord_cartesian(xlim = c(0, max(phylum$X18S.reads) + (0.1*max(phylum$X18S.reads))),
                    ylim = c(0, max(phylum$percent_ORFs) + 0.1*max(phylum$percent_ORFs))) +
  scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
  theme_classic() +
  scale_color_manual(values=P6)


ggsave(here("18S Comparison", "Class plots", "Streptophyta_Classes_18S.png"), p, width=8, height=6, dpi=300, units="in")
```

Plot individual phyla with points labeled
```{r}
data <- data_splits[["Hemichordata"]]

# with outlier point labeled by sample
p <- ggplot(data, aes(x=reads_sum, y=mgn_sum)) +
  geom_point(show.legend=TRUE, size=2, color="#0072B2") +
  geom_text(aes(label=ifelse(reads_sum>0.00002, as.character(ASV.sample),'')),
            hjust=1, vjust=-1) +
  labs(title = paste("Phylum Hemichordata in paired metagenome & 18S samples"),
         x="Fraction of COI reads", y="Fraction of eukaryotic metagenome ORFs") +
  coord_cartesian(xlim = c(0, max(data$reads_sum) + (0.1*max(data$reads_sum))),
                    ylim = c(0, max(data$mgn_sum) + 0.1*max(data$mgn_sum))) +
  scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
  theme_classic() 

#ggsave(here("18S Comparison", "Classes", "Bacillariophyta_18S_lm-slope-r2-se.png"), p,
 #     width=11, height=8, dpi=300, units="in")
```

Plot individual Class with points labeled
```{r}
phylum <- taxa[taxa$Phylum == 'Glaucocystophyceae', ]
class <- taxa[taxa$Class == 'Malacostraca', ]

# no linear regression
p <- ggplot(phylum, aes(x=COI.reads, y=count)) +
  geom_point(show.legend=TRUE, size=2) +
  labs(title = paste("Phylum Glaucocystophyceae in paired metagenome & 18S samples"),
         x="Fraction of COI reads", y="Fraction of eukaryotic metagenome ORFs") +
  coord_cartesian(xlim = c(0, max(class$X18S.reads) + (0.1*max(class$X18S.reads))),
                    ylim = c(0, max(class$count) + 0.1*max(class$count))) +
  scale_x_continuous(expand=c(0, 0)) + scale_y_continuous(expand=c(0, 0)) +
  theme_classic() 

#ggsave(here("18S Comparison", "Classes", "Class_Malacostraca_18S.png"), p,
 #     width=7, height=4, dpi=300, units="in")
```

